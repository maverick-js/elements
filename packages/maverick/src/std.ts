export * from './utils/aria';
export * from './utils/disposal';
export * from './utils/dom';
export * from './utils/event';
export * from './utils/fn';
export * from './utils/object';
export * from './utils/observable';
export * from './utils/promise';
export * from './utils/string';
export * from './utils/timing';
export * from './utils/unit';
