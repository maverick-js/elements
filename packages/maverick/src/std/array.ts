import { isArray } from './unit';

// This only exists because array.flat(Infinity) is slow.
export function flattenArray<T>(array: (T | T[])[]): T[] {
  const flat: T[] = [];

  for (let i = 0; i < array.length; i++) {
    if (isArray(array[i])) {
      flat.push(...flattenArray(array[i] as T[]));
    } else {
      flat.push(array[i] as T);
    }
  }

  return flat;
}
