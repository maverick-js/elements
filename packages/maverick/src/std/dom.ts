import type { ServerHTMLElement } from '../element/create-server-element';

export function isDOMNode(node: any): node is Node {
  return node instanceof Node;
}

export function isDOMElement(node: any): node is Element {
  return isDOMNode(node) && node.nodeType === 1;
}

export function isDOMFragment(node: any): node is DocumentFragment {
  return isDOMNode(node) && node.nodeType === 11;
}

export function createFragment(): DocumentFragment {
  return document.createDocumentFragment();
}

export function createComment(data: string): Comment {
  if (__SERVER__) return data as unknown as Comment;
  return document.createComment(data);
}

/**
 * Sets or removes the given attribute `value`. Falsy values except `''` and `0` will remove
 * the attribute. If the given `value` is a function/signal, the attribute will be updated as
 * the value updates.
 *
 * This function is safe to use on the server.
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Falsy}
 */
export function setAttribute(host: Element | ServerHTMLElement, name: string, value: unknown) {
  if (!value && value !== '' && value !== 0) {
    host.removeAttribute(name);
  } else {
    const attrValue = value + '';
    if (host.getAttribute(name) !== attrValue) {
      host.setAttribute(name, attrValue);
    }
  }
}

/**
 * Sets or removes the given style `value`. Falsy values will remove the style. If the
 * given `value` is a function/signal, the style will be updated as the value updates.
 *
 * This function is safe to use on the server.
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Falsy}
 */
export function setStyle(host: HTMLElement | ServerHTMLElement, property: string, value: unknown) {
  if (!value && value !== 0) {
    host.style.removeProperty(property);
  } else {
    host.style.setProperty(property, value + '');
  }
}

/**
 * Toggles the given class `name`. Falsy values will remove the class from the list. If the
 * given `value` is a function/signal, the class will be toggled as the value updates.
 *
 * This function is safe to use on the server.
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Falsy}
 */
export function toggleClass(host: Element | ServerHTMLElement, name: string, value: unknown) {
  host.classList[value ? 'add' : 'remove'](name);
}

/**
 * Returns elements assigned to the given slot in the shadow root. Filters out all nodes
 * which are not an element.
 *
 * @param el - The element containing the slot.
 * @param name - The name of the slot (optional).
 */
export function getSlottedChildren(el: HTMLElement, name?: string): Element[] {
  const selector = name ? `slot[name="${name}"]` : 'slot:not([name])';
  const slot = el.shadowRoot?.querySelector(selector) as HTMLSlotElement | null;
  const childNodes = slot?.assignedNodes({ flatten: true }) ?? [];
  return Array.prototype.filter.call(childNodes, (node) => node.nodeType == 1);
}

export function attachDeclarativeShadowDOM(element: HTMLElement) {
  const template = element.firstChild as HTMLTemplateElement;
  const mode = template.getAttribute('shadowroot')! as 'open' | 'closed';
  const shadowRoot = (template.parentNode as HTMLElement).attachShadow({ mode });
  shadowRoot.appendChild(template.content);
  template.remove();
}
