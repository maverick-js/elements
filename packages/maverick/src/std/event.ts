import type { Constructor } from 'type-fest';

import { type Dispose, type JSX, onDispose, untrack } from '../runtime';
import { noop } from './unit';

const DOM_EVENT = Symbol('DOM_EVENT'),
  DOMEventBase: Constructor<Event> = __SERVER__ ? (class Event {} as any) : Event;

export interface DOMEventInit<Detail = unknown> extends EventInit {
  readonly detail: Detail;
  readonly triggerEvent?: Event;
}

export class DOMEvent<Detail = unknown> extends DOMEventBase {
  /**
   * The event detail.
   */
  readonly detail!: Detail;

  /**
   * The preceding event that was responsible for this event being fired.
   */
  readonly triggerEvent?: Event;

  /**
   * Walks up the event chain (following each `triggerEvent`) and returns the origin event
   * that started the chain.
   */
  readonly originEvent!: Event;

  /**
   * Walks up the event chain (following each `triggerEvent`) and determines whether the initial
   * event was triggered by the end user (ie: check whether `isTrusted` on the `originEvent` `true`).
   */
  readonly isOriginTrusted!: boolean;

  constructor(type: string, init?: DOMEventInit<Detail>) {
    super(type, init);

    Object.defineProperty(this, DOM_EVENT, {
      get: () => true,
    });

    defineEventProperty(this, 'detail', () => init?.detail);
    defineEventProperty(this, 'triggerEvent', () => init?.triggerEvent);
    defineEventProperty(this, 'originEvent', () => getOriginEvent(this) ?? this);
    defineEventProperty(this, 'isOriginTrusted', () => getOriginEvent(this)?.isTrusted ?? false);
  }
}

/**
 * Creates and returns a `DOMEvent`. This function is typed to match all events declared
 * on the global `HTMLElementEventMap`. You can extend it like so:
 *
 * ```ts
 * declare global {
 *   interface HTMLElementEventMap {
 *     foo: DOMEvent<number>;
 *   }
 * }
 * ```
 */
export function createEvent<EventType extends keyof HTMLElementEventMap>(
  type: EventType,
  ...init: InferEventDetail<EventType> extends void | undefined | never
    ? [init?: Partial<InferEventInit<EventType>>]
    : [init: InferEventInit<EventType>]
): EventType extends keyof HTMLElementEventMap
  ? HTMLElementEventMap[EventType] extends DOMEvent
    ? HTMLElementEventMap[EventType]
    : DOMEvent<InferEventDetail<EventType>>
  : DOMEvent<InferEventDetail<EventType>> {
  return new DOMEvent(type, init[0] as any) as any;
}

/**
 * Creates a `DOMEvent` and dispatches it from the given `target`. This function is typed to
 * match all events declared on the global `HTMLElementEventMap`. You can extend it like so:
 *
 * ```ts
 * declare global {
 *   interface HTMLElementEventMap {
 *     foo: DOMEvent<number>;
 *   }
 * }
 * ```
 */
export function dispatchEvent<EventType extends keyof HTMLElementEventMap>(
  target: EventTarget | null,
  event: EventType,
  ...init: InferEventDetail<EventType> extends void | undefined | never
    ? [init?: Partial<InferEventInit<EventType>>]
    : [init: InferEventInit<EventType>]
): boolean {
  if (__SERVER__) return false;
  return target ? target.dispatchEvent(new DOMEvent(event, init[0] as any)) : false;
}

/**
 * Whether the given `event` is a Maverick DOM Event class.
 */
export function isDOMEvent(event?: Event | null): event is DOMEvent<unknown> {
  return !!event?.[DOM_EVENT];
}

/**
 * Whether the given `event` is of the given `type`.
 */
export function isEventType<Type extends keyof HTMLElementEventMap>(
  event: Event,
  type: Type,
): event is HTMLElementEventMap[Type] {
  return event.type === type;
}

/**
 * Walks up the event chain (following each `triggerEvent`) and returns the origin event that
 * started the chain.
 */
export function getOriginEvent(event: DOMEvent): Event | undefined {
  let triggerEvent = event.triggerEvent as DOMEvent;

  while (triggerEvent && triggerEvent.triggerEvent) {
    triggerEvent = triggerEvent.triggerEvent as DOMEvent;
  }

  return triggerEvent;
}

/**
 * Walks an event chain on a given `event`, and invokes the given `callback` for each trigger event.
 *
 * @param event - The event on which to follow the chain.
 * @param callback - Invoked for each trigger event in the chain. If a `value` is returned by
 * this callback, the walk will end and `[event, value]` will be returned.
 */
export function walkTriggerEventChain<T>(
  event: Event,
  callback: (event: Event) => NonNullable<T> | void,
): [event: Event, value: NonNullable<T>] | undefined {
  if (!isDOMEvent(event)) return;

  let triggerEvent = event.triggerEvent as DOMEvent;

  while (triggerEvent) {
    const returnValue = callback(triggerEvent);
    if (returnValue) return [triggerEvent, returnValue];
    triggerEvent = triggerEvent.triggerEvent as DOMEvent;
  }

  return;
}

/**
 * Attempts to find a trigger event with a given `eventType` on the event chain.
 *
 * @param event - The event on which to look for a trigger event.
 * @param type - The type of event to find.
 */
export function findTriggerEvent<Type extends string>(
  event: Event,
  type: Type,
): (Type extends keyof HTMLElementEventMap ? HTMLElementEventMap[Type] : Event) | undefined {
  return walkTriggerEventChain(event, (e) => e.type === type)?.[0] as any;
}

/**
 * Whether a trigger event with the given `eventType` exists can be found in the event chain.
 *
 * @param event - The event on which to look for a trigger event.
 * @param type - The type of event to find.
 */
export function hasTriggerEvent(event: Event, type: string): boolean {
  return !!findTriggerEvent(event, type);
}

/**
 * Appends the given `triggerEvent` to the event chain. This means the new origin event will be
 * the origin of the given `triggerEvent`, or the `triggerEvent` itself (if no chain exists on the
 * trigger).
 *
 * @param event - The event on which to extend the trigger event chain.
 * @param triggerEvent - The trigger event that will becoming the new origin event.
 */
export function appendTriggerEvent(event: DOMEvent, triggerEvent?: Event) {
  const origin = (getOriginEvent(event) ?? event) as DOMEvent;
  defineEventProperty(origin, 'triggerEvent', () => triggerEvent);
}

function defineEventProperty<T extends keyof DOMEvent>(
  event: DOMEvent,
  prop: T,
  value: () => DOMEvent[T],
) {
  Object.defineProperty(event, prop, {
    enumerable: true,
    configurable: true,
    get: value,
  });
}

export type InferEventDetail<T> = T extends string
  ? T extends keyof HTMLElementEventMap
    ? HTMLElementEventMap[T] extends DOMEvent<infer Detail>
      ? Detail
      : unknown
    : unknown
  : T extends DOMEventInit<infer Detail>
  ? Detail
  : T extends DOMEvent<infer Detail>
  ? Detail
  : T extends CustomEvent<infer Detail>
  ? Detail
  : unknown;

export type InferEventInit<T> = T extends string
  ? T extends keyof HTMLElementEventMap
    ? InferEventInit<HTMLElementEventMap[T]>
    : DOMEventInit<unknown>
  : T extends Constructor<DOMEvent>
  ? DOMEventInit<InferEventDetail<InstanceType<T>>>
  : T extends DOMEvent
  ? DOMEventInit<InferEventDetail<T>>
  : T extends DOMEventInit
  ? T
  : DOMEventInit<unknown>;

/**
 * Adds an event listener for the given `type` and returns a function which can be invoked to
 * remove the event listener.
 *
 * - The listener is removed if the current scope is disposed.
 * - This function is safe to use on the server (noop).
 */
export function listenEvent<Target extends EventTarget, EventType extends string>(
  target: Target,
  type: EventType,
  handler: JSX.TargetedEventHandler<
    Target,
    EventType extends keyof HTMLElementEventMap ? HTMLElementEventMap[EventType] : Event
  >,
  options?: AddEventListenerOptions | boolean,
): Dispose {
  if (__SERVER__) return noop;
  target.addEventListener(type, handler as any, options);
  return onDispose(() => target.removeEventListener(type, handler as any, options));
}

export function isPointerEvent(event: Event | undefined): event is PointerEvent {
  return !!event?.type.startsWith('pointer');
}

export function isTouchEvent(event: Event | undefined): event is TouchEvent {
  return !!event?.type.startsWith('touch');
}

export function isMouseEvent(event: Event | undefined): event is MouseEvent {
  return /^(click|mouse)/.test(event?.type ?? '');
}

export function isKeyboardEvent(event: Event | undefined): event is KeyboardEvent {
  return !!event?.type.startsWith('key');
}

export function wasEnterKeyPressed(event: Event | undefined) {
  return isKeyboardEvent(event) && event.key === 'Enter';
}

export function wasEscapeKeyPressed(event: Event | undefined) {
  return isKeyboardEvent(event) && event.key === 'Escape';
}

export function isKeyboardClick(event: Event | undefined) {
  return isKeyboardEvent(event) && (event.key === 'Enter' || event.key === ' ');
}
