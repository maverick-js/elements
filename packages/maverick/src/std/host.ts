import { getElementInstance } from '../element/internal';
import type { AnyMaverickElement } from '../element/types';

export function useHost<Element extends AnyMaverickElement>(): Element {
  const instance = getElementInstance();

  if (!instance) {
    throw Error(__DEV__ ? '[maverick] called `useHost` outside of root or setup function' : '');
  }

  return instance.host as unknown as Element;
}
