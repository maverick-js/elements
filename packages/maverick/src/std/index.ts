export * from './aria';
export * from './disposal';
export * from './dom';
export * from './event';
export * from './fn';
export * from './host';
export * from './object';
export * from './observable';
export * from './promise';
export * from './string';
export * from './timing';
export * from './unit';
