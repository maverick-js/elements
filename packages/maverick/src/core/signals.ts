import { effect as $effect, type Effect, type StopEffect } from '@maverick-js/signals';

import { noop } from '../std/unit';

export {
  type Callable,
  type Computation,
  type ComputedSignalOptions,
  type Disposable,
  type Dispose,
  type Effect,
  type InferSignalValue,
  type Maybe,
  type MaybeDisposable,
  type MaybeFunction,
  type MaybeSignal,
  type MaybeStopEffect,
  type NextValue,
  type ReadSignal,
  type Scope,
  type SignalOptions,
  type StopEffect,
  type WriteSignal,
  createScope,
  computed,
  getScope,
  isReadSignal,
  isWriteSignal,
  onDispose,
  onError,
  peek,
  readonly,
  root,
  scoped,
  signal,
  tick,
  untrack,
} from '@maverick-js/signals';

/**
 * Invokes the given function each time any of the signals that are read inside are updated
 * (i.e., their value changes). The effect is immediately invoked on initialization.
 *
 * @see {@link https://github.com/maverick-js/signals#effect}
 */
export const effect = (__SERVER__ ? serverEffect : $effect) as typeof $effect;

function serverEffect(effect: Effect, options?: { id?: string }): StopEffect {
  if (typeof process !== 'undefined' && process.env?.NODE_ENV === 'test') {
    return $effect(effect, options);
  }

  return noop;
}
