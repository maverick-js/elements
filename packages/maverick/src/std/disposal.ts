import { onDispose } from '../core/signals';

export interface DisposalBin {
  add(...callbacks: (() => any)[]): void;
  empty(): void;
}

/**
 * Creates and returns a disposal bin for managing cleanup operations.
 *
 * @example
 * ```ts
 * const bin = createDisposalBin();
 * // Add callbacks
 * bin.add(() => {});
 * bin.add(() => {});
 * // Empty bin by running callbacks
 * bin.empty();
 * ```
 */
export function createDisposalBin(): DisposalBin {
  const disposal = new Set<() => unknown>();
  return {
    add(...callbacks) {
      for (const callback of callbacks) disposal.add(callback);
    },
    empty() {
      for (const callback of disposal) callback();
      disposal.clear();
    },
  };
}

/**
 * Creates and return a `DisposalBin`. This bin is emptied if the parent scope is disposed of.
 */
export function useDisposalBin() {
  const disposal = createDisposalBin();
  onDispose(disposal.empty);
  return disposal;
}
