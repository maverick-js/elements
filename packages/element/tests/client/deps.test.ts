import {
  defineMaverickElement,
  type MaverickElement,
  SETUP_STATE_SYMBOL,
} from '@maverick-js/element';
import { getContext, setContext } from '@maverick-js/signals';
import { waitAnimationFrame, waitTimeout } from '@maverick-js/std';
import {
  createContext,
  type CustomElementOptions,
  MaverickComponent,
  onConnect,
  onError,
  onSetup,
  provideContext,
  useContext,
} from 'maverick.js';

const target = document.body;

afterEach(() => {
  target.innerHTML = '';
});

it('should wait for parents to connect', async () => {
  const Context = createContext<number[]>(() => []);

  const error = new Error(),
    errorHandler = vi.fn();

  class ParentA extends MaverickComponent {
    static element: CustomElementOptions = {
      name: 'mk-parent-a',
    };

    constructor() {
      super();
      setContext('foo', 10);
      provideContext(Context, [1]);
      onError(errorHandler);
    }
  }

  class ParentB extends MaverickComponent {
    static element: CustomElementOptions = {
      name: 'mk-parent-b',
    };

    constructor() {
      super();
      onSetup(this.#onSetup.bind(this));
    }

    #onSetup() {
      const value = useContext(Context);
      expect(value).toEqual([1]);
      provideContext(Context, [...value, 2]);
    }
  }

  class Child extends MaverickComponent {
    static element: CustomElementOptions = {
      name: 'mk-child',
    };

    constructor() {
      super();
      onSetup(this.#onSetup.bind(this));
    }

    #onSetup() {
      expect(getContext('foo')).toBe(10);
      const value = useContext(Context);
      expect(value).toEqual([1, 2]);
      provideContext(Context, [...value, 3]);
    }
  }

  class GrandChild extends MaverickComponent {
    static element: CustomElementOptions = {
      name: 'mk-grandchild',
    };

    constructor() {
      super();
      onSetup(this.#onSetup.bind(this));
      onConnect(this.#onConnect.bind(this));
    }

    #onSetup() {
      const value = useContext(Context);
      expect(value).toEqual([1, 2, 3]);
    }

    #onConnect() {
      throw error;
    }
  }

  const parentA = document.createElement(ParentA.element.name);

  const parentB = document.createElement(ParentB.element.name);
  parentA.append(parentB);

  const child = document.createElement(Child.element.name);
  parentB.append(child);

  const grandchild = document.createElement(GrandChild.element.name);
  child.append(grandchild);

  target.append(parentA);

  expect(target).toMatchSnapshot();

  defineMaverickElement(GrandChild);
  expect(grandchild[SETUP_STATE_SYMBOL] === 1).toBeTruthy();

  defineMaverickElement(Child);
  expect(child[SETUP_STATE_SYMBOL] === 1).toBeTruthy();

  window.customElements.whenDefined(GrandChild.element.name);
  window.customElements.whenDefined(Child.element.name);

  expect(child[SETUP_STATE_SYMBOL] === 1).toBeTruthy();
  expect(grandchild[SETUP_STATE_SYMBOL] === 1).toBeTruthy();

  defineMaverickElement(ParentB);
  window.customElements.whenDefined(ParentB.element.name);

  expect(parentB[SETUP_STATE_SYMBOL] === 1).toBeTruthy();

  defineMaverickElement(ParentA);
  window.customElements.whenDefined(ParentA.element.name);

  expect(parentA[SETUP_STATE_SYMBOL] === 2).toBeTruthy();

  await waitAnimationFrame();

  expect(parentB[SETUP_STATE_SYMBOL] === 2).toBeTruthy();
  expect(child[SETUP_STATE_SYMBOL] === 2).toBeTruthy();
  expect(grandchild[SETUP_STATE_SYMBOL] === 2).toBeTruthy();

  parentA.remove();

  await waitTimeout(0);
  await waitAnimationFrame();

  expect((parentA as MaverickElement).$).toBeUndefined();
  expect((parentB as MaverickElement).$).toBeUndefined();
  expect((child as MaverickElement).$).toBeUndefined();
  expect((grandchild as MaverickElement).$).toBeUndefined();

  expect(errorHandler).toBeCalledTimes(1);
  expect(errorHandler).toHaveBeenCalledWith(error);
});
