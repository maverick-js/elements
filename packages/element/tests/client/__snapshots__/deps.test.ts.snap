// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should wait for parents to connect 1`] = `
<body>
  
    
  
  


  <mk-parent-a>
    <mk-parent-b>
      <mk-child>
        <mk-grandchild />
      </mk-child>
    </mk-parent-b>
  </mk-parent-a>
</body>
`;
